lib/SILOptimizer/LoopTransforms/ArrayPropertyOpt.cpp

```diff
@@ -107,26 +150,55 @@ class ArrayPropertiesAnalysis {
     // beneficial. This heuristic also simplifies which regions we want to
     // specialize on. We will specialize the outermost loopnest that has
     // 'array.props' instructions in its preheader.
+
     bool FoundHoistable = false;
     for (auto *BB : Loop->getBlocks()) {
       for (auto &Inst : *BB) {
-
         // Can't clone alloc_stack instructions whose dealloc_stack is outside
         // the loop.
-        if (!Loop->canDuplicate(&Inst))
+        if (!Loop->canDuplicate(&Inst)) {
           return false;
+        }
 
         ArraySemanticsCall ArrayPropsInst(&Inst, "array.props", true);
         if (!ArrayPropsInst)
           continue;
 
-        if (!canHoistArrayPropsInst(ArrayPropsInst))
+        if (!canHoistArrayPropsInst(ArrayPropsInst)) {
           return false;
+        }
+
         FoundHoistable = true;
       }
     }
 
-    return FoundHoistable;
+    if (!FoundHoistable) {
+      return false;
+    }
+
+    // Additionally, we don't specialize the loop if the InstCount exceeds a
+    // predefined threshold or we find opaque calls.
+    // Since only few loops qualify as hoistable, and the profitability check
+    // can run long in cases of large thresholds, these checks are not folded
+    // along with the legality checks above.
+
+    uint32_t InstCount = 0;
+
+    for (auto *BB : Loop->getBlocks()) {
+      for (auto &Inst : *BB) {
+        if (InstCount++ > MaxInstThreshold) {
+          return false;
+        }
+
+        if (auto Apply = FullApplySite::isa(&Inst)) {
+          const auto Callee = Apply.getReferencedFunctionOrNull();
+          auto CalleeInstCount = checkProfitabilityRecursively(Callee);
```

Comment:
I'm a bit concerned about the complexity. You're scanning a potentially large part of the whole module for _every_ function on which this pass runs. This is quadratic.
But it might be no problem if you reduce MaxInstThreshold.
Alternatively, you could make a simple Analysis, which contains the InstCountCache. This would cache the instruction count for functions across multiple calls of this pass.

Reply:
Ok. I was worried about the cache not being fully utilized as well. I'll see how this can be improved. Thanks for the suggestion.

Reply:
@meg-gupta @eeckstein 
Here are some suggestions:

We could have a general bottom-up inter-procedural analysis for "estimated function computation complexity". It would still need a threshold though, or the analysis will be much more expensive than what you're already doing. The cost to compute it is `num_functions * threshold`, where the threshold is reached instantly at any external, virtual, or witness call (it should not follow all known callees). That should be very cheap in practice. But, we should also have a multiplier for the level of the instruction in the loop nest, which means computing loops.

I think this is overkill for just this one purpose, but it seems like a generally useful thing to have. We should find out if others agree before spending time on it.

Just to kill this argument before someone brings it up... I do *not* think this should be combined with the algorithm for inlining cost analysis. None of the inliner's complexity is relevant to this metric. It's designed around processing blocks and instructions in a particular order to propagate constants. And it's determining code size cost, not computation complexity. Reusing the logic for unrelated metrics would be confusing. However, `instructionInlineCost` is currently simple enough that it could be reused for both code size and computation cost of a single instruction at a time. It essentially allows you to ignore most SIL instructions as having no impact on generated code.

Reply:
Yeah, a general bottom-up analysis sounds good for this purpose.
And sorry, I have to bring this up: the ShortestPathAnalysis of the inliner serves a very similar purpose. So it would be cool if we could extract the ShortestPathAnalysis from the inliner and make it a general tool which can also be used here.